#issue-13: View details of the list items

hajarNasr opened the pull request:
## Type of Changes

|      | Type                       |
| --- | -------------------------- |
|  ✓    | :sparkles: New feature     |
|  ✓    | :bug: Bug fix              |

Closes #13 

## Acceptance Criteria
- [X]  From the shopping list, the user can tap to navigate to a detailed view of the item
- [X] Detail view includes the following information:
      * Name of the item
      * Last purchased date
      * Estimated number of days until the next purchase
      * Number of times the item has been purchased
- [X]  When in the “item details” view, a back arrow should appear in the header that takes the user back to the “list” view.

## Description
**This PR adds a details-view to the list items and fixes semantic errors in the previous code**

**First:** It creates `ListItemDetails` which is a component responsible for showing the user the details of an item once the user clicks the info button associated with that item.

**Second:** It adds three new functions to `helper.js`:
   * `getDaysSinceLastPurchase`: Gets the number of days since the last purchase.
   * `calculateNextPurchase`: Calculates the estimated date for the next purchase.
  * `getDaysUntilNextPurchase`: Gets the number of estimated days until the next purchase.

**Third**: It fixes the way the items are colored and sorted on the list.

## Update
A user can now view the details of the items on their list.
## How things look so far
![the details-view of the list items](https://i.ibb.co/qxKYB85/detailsview.gif)

## Testing Steps / QA Criteria

* From your terminal, pull down this branch with `git pull origin se-hn-issue-13` 
* Check that branch out with `git checkout hn-yk-issue-13`
* Run `yarn` and `yarn start` to run the [local server](http://localhost:3000/).
* Join a list by using the **"wool haifa muong"** token, or create your own list and add some items to it.
* Click the info button to an item right next to the trash icon and a modal should pop up to show you the details of the item. If you've never purchased that item before a **No recorded purchases yet** message should show. Otherwise, some info about that item should show.  

**Note**: In case you needed to, please use [this converter](https://datetomilliseconds.netlify.app/) to convert dates to milliseconds.

**Thank you so much for your time.**

Commit:
feat: create ListItemDetails.js and display it as a modal

saraheisa committed:
Feat: Add item details to modal

Format dates to mon day, year
Calculate next purchase date

saraheisa committed:
Style: Adjust app width and positon

- Made the app take full width

saraheisa committed:
Feat: Close the modal by clicking anywhere in the window

- Added a container for the modal with full width&height
to close the modal whenever you click on the window

saraheisa committed:
Style: add a funky animation to the modal

Commit:
style: add a little CSS to the modal close button

Commit:
feat: add a few aria labels for ally

Commit:
style: add info fa icon to the details btn

Commit:
fix: sort the items based on how many days until the next purchase

saraheisa committed:
Fix Typo

Commit:
style: add a few media queries

deboragaleano commented on src/components/ListItemDetails.js:
```diff
@@ -0,0 +1,47 @@
+const ListItemDetails = ({ listItem, isShowModal, hideModal }) => {
+  const toggleModal = isShowModal ? 'display-block' : 'display-none';
+  return (
+    <div className={`modal ${toggleModal}`} onClick={hideModal}>
```
I love this modal! It's so cool 💯

deboragaleano commented on src/lib/helper.js:
```diff
@@ -36,15 +36,41 @@ const fromMilliSecToHours = (time) => time / hourInMilliSeconds;
-  fromMilliSecToDays(getUTCNowInMilliSec()) -
-    fromMilliSecToDays(recentPurchase) >
-    howSoon * 2;
+export const isOutOfDate = ({ howSoon, recentPurchase }) => {
```
wow, great work! Very clever!  💡 👏

segdeha replied:
Agreed! This is great!

deboragaleano reviewed (commented):
This was a lot of work it seems! I'm so impressed (especially with the helper functions!) 👏 
Everything works perfectly 👌

segdeha commented on src/components/ListItem.js:
```diff
@@ -57,14 +59,20 @@ const ListItem = ({ listItem, itemId }) => {
+
+    // if the current item is not out of date get the number of days
+    // until the next purchase
+    const daysUntilNextPurchase = getDaysUntilNextPurchase(listItem);
```
Love these little, single-responsibility functions! They're great candidates for unit tests!

segdeha commented on src/components/ListItem.js:
```diff
@@ -57,14 +59,20 @@ const ListItem = ({ listItem, itemId }) => {
+    // return the status and ariaLabel of the item based on that number
     switch (true) {
-      case howSoon <= 7:
+      case daysUntilNextPurchase <= 7:
```
Changing this variable name is a big improvement!

segdeha commented on src/components/ListItem.js:
```diff
@@ -77,7 +85,10 @@ const ListItem = ({ listItem, itemId }) => {
+        aria-label={`${listItem.name} ${ariaLabel} ${
+          isChecked && 'Already purchased today'
+        }`}
+        tabIndex="0"
```
Curious what you’re trying to accomplish with `tabIndex` here? Best to avoid setting it manually unless you really need to

saraheisa replied:
We've used [ChromeVox](https://chrome.google.com/webstore/detail/chromevox-classic-extensi/kgejglhpjiefppelpmljglcjbhoiplfn) to test accessibility and it ignores this content without ` tabindex ` that's why we used it

segdeha commented on src/components/ListItem.js:
```diff
@@ -87,8 +98,22 @@ const ListItem = ({ listItem, itemId }) => {
         disabled={isChecked}
         checked={isChecked}
-        aria-label={isChecked ? 'Purchased item' : 'Check to mark as purchased'}
+        aria-label="Check to mark as purchased"
```
I’m trying to run the code in my head to understand what the experience is like on a screen reader when an item is disabled. Have you tried it in VoiceOver or NVDA? It seems like the user might get conflicting messages ("Already purchased today" followed by "Check to mark as purchased"), no?

saraheisa replied:
In `ChromeVox` the screenreader ignored the disabled checkboxes

llunaplanet commented on src/components/ListItemDetails.js:
```diff
@@ -0,0 +1,47 @@
+              <p>
+                You have purchased this item {listItem['numberOfPurchases']}{' '}
+                time
+                {listItem['numberOfPurchases'] > 1 && 's'}.
```
Cool detail taking care of singular and plural forms 👍

segdeha replied:
Agreed! Nice attention to detail!

llunaplanet reviewed (approved):
All acceptance criteria has been meet, great work! 👍

segdeha commented on src/components/ListItemDetails.js:
```diff
@@ -0,0 +1,47 @@
+  return (
+    <div className={`modal ${toggleModal}`} onClick={hideModal}>
+      <div className="modal-container">
+        <div className="modal-header" tabIndex="0">
```
Similar to my comment on `ListItem.js`, I’m curious what problem you’re trying to solve by setting the `tabIndex` manually?

segdeha commented on src/components/ListItemDetails.js:
```diff
@@ -0,0 +1,47 @@
+        </div>
+        <div className="modal-body" tabIndex="0">
+          {listItem.recentPurchase ? (
+            <>
```
Do you need the fragment here? I’d have thought React would be happy with the structure without it.

saraheisa replied:
the app is throwing a parsing error  `Parsing error: Adjacent JSX elements must be wrapped in an enclosing tag`

segdeha commented on src/components/SortedList.js:
```diff
@@ -21,16 +25,20 @@ const SortedList = ({ itemsKeys, data }) => {
+          // the list renders
           let [howSoonA, nameA, howSoonB, nameB] = [
-            data[a].howSoon,
+            Math.abs(getDaysUntilNextPurchase(data[a])),
```
Help me understand why you need `Math.abs` here. I’m guessing you found situations where the date when we predicted the item would be bought has passed and this is resulting in a negative value being returned from `getDaysUntilNextPurchase`. Is that right?

If my assumption above is correct, wouldn’t you want to sort on the true value and not the non-negative value? I.e. if the days until next purchase is -14, using `Math.abs` would yield 14 which would put the item in the “kinda soon” group, no?

I’m pretty sure you are doing the right thing and I’m just not following, but thanks for entertaining my question! 🙂

hajarNasr replied:
Hey, @segdeha! Thanks for this great review. :heart:

Yes, your assumption is correct. 
When an item is out of date, the days until the next purchase will be a negative value. 

However, we only use these values/days `howSoonA` and `howSoonB` when we call `compareItems()`, and we call that function only if both items are out of date or if both items are not. 

When we compare two outdated items both values will be negative. So we use `Math.abs` to make sure that the item with more elapsed days comes second. The item with -14 days until the next purchase should be greater than the one with -10 days.

I think it's safe to use `Math.abs` here because we're always going to compare between two negative values or two positive values. I'll try to add a comment here to make things more clear. Thank you again. :)

segdeha replied:
Ah, OK, that makes sense! Thank you!

segdeha commented on src/css/components/ItemsList.css:
```diff
@@ -80,11 +80,20 @@
+.item-details-icon {
+  order: 3;
+  border: none;
+  color: rgb(80, 79, 79);
```
Gray with just the _slightest_ tinge of red? 😅

saraheisa replied:
we're so precise you can't tell 😂
.
.
.

actually we just select a color from the color picker and don't look at the numbers

segdeha commented on src/css/components/ItemsList.css:
```diff
@@ -80,11 +80,20 @@
   border: none;
   font-size: 1em;
-  order: 3;
+  order: 4;
```
Oh wow, today I learned! I didn’t know about the `order` property!

segdeha commented on src/css/components/Navigation.css:
```diff
@@ -1,18 +1,22 @@
+.navbar {
+  position: absolute;
+  bottom: 0px;
```
When you have `0` for the value, you can always omit the unit (0 of any unit is just...0). So this can change to `bottom: 0;`

segdeha commented on src/css/components/main.css:
```diff
@@ -5,14 +5,27 @@
+}
+@media (max-width: 550px) {
+  .App {
+    width: 95%;
```
All of these percentages feel like “magic” numbers. Consider moving them into semantically-named CSS variables such as `--tablet-width`

segdeha commented on src/css/components/modal.css:
```diff
@@ -0,0 +1,71 @@
+  width: 100%;
+  height: 100%;
+  top: 0;
+  background-color: #00000059;
```
Wow, y’all are teaching me all kinds of things in this PR! 😅 When I learned CSS, you couldn’t specify the alpha channel in hex colors, so I learned to use the `rgba()` function instead. So, I never learned that this 8-digit syntax was supported until right now! Thank you!

segdeha commented on src/css/components/modal.css:
```diff
@@ -0,0 +1,71 @@
+  height: 100%;
+  top: 0;
+  background-color: #00000059;
+  z-index: 99999;
```
`z-index` values are relative to the [stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context) of the elements, so `1` probably would work here just as well as `99999` 🙂

segdeha commented on src/css/components/modal.css:
```diff
@@ -0,0 +1,71 @@
+  color: #c04239;
+}
+/* Add Animation */
+@-webkit-keyframes animatetop {
```
Gold star for completeness 🙂 but the browsers that don’t support the standard `@keyframes` syntax are unlikely to be able to run this app, anyway, so you could probably get rid of this fallback code.

saraheisa replied:
Yup noted!

segdeha commented on src/css/components/modal.css:
```diff
@@ -0,0 +1,71 @@
+  }
+}
+
+@keyframes animatetop {
```
Love that you took the care to animate in your modal!!!

Not blocking, but something to consider is that animating a `translate3d` can result in smoother tweens than animating properties such as `top` or `left`. It has to do with how the browser renders the elements. Any of the 3D functions (`translate3d`, `translateZ`, and strangely enough `opacity`) cause the animation to be offloaded to the computer’s GPU rather than relying on the CPU to do the necessary calculations. It’s usually _fine_ on modern browsers + capable devices, but if you have users on older hardware it could make a noticeable difference.

saraheisa replied:
yeah 3d will be smoother

segdeha reviewed (approved):
This looks and works great!!! No blocking feedback from me, but I have a few questions and a couple of things for consideration. Nice job!!!

saraheisa committed:
Style: Change some modal Styles

    - Remove webkit animation fallback because it's not necessary as the app
    won't run on that kind of browsers anyway
    - Added a bounce effect to the animation
    - change z-index to 1 because why not
    - use hex format for all the colors

saraheisa committed:
Style: Change some list styles
     - Standardize colors to use hex format
     - Remove repeated styles

saraheisa committed:
Style: use CSS vars for media quieries width to add semantic meaning

saraheisa committed:
Style: Truncate px from 0

because zeros are self-sufficient!!!

saraheisa committed:
Merge branch 'main' into se-hn-issue-13

saraheisa committed:
Fix: Resolve conflicts
